communities/src/lib.rs

```diff
@@ -67,58 +75,77 @@ decl_module! {
+            nominal_income: Option<NominalIncomeType>
+        ) {
             debug::RuntimeLogger::init();
             let sender = ensure_signed(origin)?;
+            if let Some(d) = demurrage {
+                validate_demurrage(&d).map_err(|_| <Error<T>>::InvalidDemurrage)?;
+            }
+            if let Some(i) = nominal_income {
+                validate_nominal_income(&i).map_err(|_| <Error<T>>::InvalidNominalIncome)?;
+            }
+
             let cid = CommunityIdentifier::from(blake2_256(&(loc.clone(), bootstrappers.clone()).encode()));
             let cids = Self::community_identifiers();
             ensure!(!cids.contains(&cid), "community already registered");
+            Self::validate_locations(&loc, &cids)?; // <- O(n^2) !!!
 
-            for l1 in loc.iter() {
-                ensure!(Self::is_valid_geolocation(&l1), "invalid geolocation specified");
-                //test within this communities' set
-                for l2 in loc.iter() {
-                    if l2 == l1 { continue }
-                    ensure!(Self::solar_trip_time(&l1, &l2) >= MIN_SOLAR_TRIP_TIME_S, "minimum solar trip time violated within supplied locations");
-                }
-                // prohibit proximity to poles
-                if Self::haversine_distance(&l1, &NORTH_POLE) < DATELINE_DISTANCE_M
-                    || Self::haversine_distance(&l1, &SOUTH_POLE) < DATELINE_DISTANCE_M {
-                    debug::warn!(target: LOG, "location too close to pole: {:?}", l1);
-                    return Err(<Error<T>>::MinimumDistanceViolationToPole.into());
-                }
-                // prohibit proximity to dateline
-                let dateline_proxy = Location { lat: l1.lat, lon: DATELINE_LON };
-                if Self::haversine_distance(&l1, &dateline_proxy) < DATELINE_DISTANCE_M {
-                    debug::warn!(target: LOG, "location too close to dateline: {:?}", l1);
-                    return Err(<Error<T>>::MinimumDistanceViolationToDateLine.into());
-                }
-                // test against all other communities globally
-                for other in cids.iter() {
-                    for l2 in Self::locations(other) {
-                        if Self::solar_trip_time(&l1, &l2) < MIN_SOLAR_TRIP_TIME_S {
-                            debug::warn!(target: LOG,
-                                "location {:?} too close to previously registered location {:?} with cid {:?}",
-                                l1, l2, other);
-                            return Err(<Error<T>>::MinimumDistanceViolationToOtherCommunity.into());
-                        }
-                    }
-                }
-            }
+            // All checks done, now mutate state
 
             <CommunityIdentifiers>::mutate(|v| v.push(cid));
             <Locations>::insert(&cid, &loc);
             <Bootstrappers<T>>::insert(&cid, &bootstrappers);
+            // Todo: Validate metadata??
```

Comment:
yes!
* validate that contents is utf8 if need be
* limit length of PalletStrings (at beginning of fn, not at the end)

Reply:
> For future PR's: Please try keeping them small and make one PR per change to be reviewed. Here I had to review test improvements and cosmetics along with the actual task, which makes it confusing

Yes sorry about that. I agree and I should have taken the extra effort. It kind of emerged organically, as I had to use the `register_test_currency` in a new place and thought that instead of copy-pasting it to yet another location I solve the issue. But I will be extra strict with myself the next time. :)

Reply:
Found a `no_std` compatible utf8 encoding lib [encoiding_rs](https://github.com/hsivonen/encoding_rs)

Will add validation with it.